eclass/zig-toolchain.eclass

```diff
@@ -0,0 +1,382 @@
+				only | yes) features+="+soft_float";;
+				softfp | no) features+="-soft_float";;
+				*) die "${FUNCNAME[0]}: tc-tuple-is-softfloat returned unexpected value"
+			esac
+		;;
+	esac
+
+	echo ${base_cpu}${features}
+}
+
+# @FUNCTION: zig-toolchain_find_installation
+# @DESCRIPTION:
+# Detects suitable Zig installation and sets ZIG_VER and ZIG_EXE
+# variables.
+#
+# See ZIG_EXE and ZIG_VER descriptions for more information.
+zig-toolchain_find_installation() {
+	# Adapted from https://github.com/gentoo/gentoo/pull/28986
+	# Many thanks to Florian Schmaus (Flowdalic)!
+
+	[[ -n ${ZIG_SLOT} ]] || die "${FUNCNAME[0]}: ZIG_SLOT must be set"
+	if ver_test "${ZIG_SLOT}" -lt "0.13"; then
+		die "${ECLASS}: ZIG_SLOT must be >= 0.13, found ${ZIG_SLOT}"
+	fi
+
+	einfo "Searching Zig ${ZIG_SLOT}..."
+
+	local zig_supported_versions=(
+		"9999"
+		"0.13.1"
+		"0.13.0"
+	)
+
+	local base_path="${BROOT}/usr/bin"
+
+	local selected_path selected_ver
+	for selected_ver in "${zig_supported_versions[@]}"; do
+		# Check if candidate satisfies ZIG_SLOT condition.
+		local candidate_slot
+		candidate_slot=$(ver_cut 1-2 ${selected_ver})
+		if ver_test "${candidate_slot}" -ne ${ZIG_SLOT}; then
+			continue
+		fi
+
+		local candidate_path
+		# Prefer "dev-lang/zig" over "dev-lang/zig-bin"
+		candidate_path="${base_path}/zig-${selected_ver}"
+		if [[ -x "${candidate_path}" ]]; then
+			selected_path="${candidate_path}"
+			break;
+		fi
+
+		candidate_path="${base_path}/zig-bin-${selected_ver}"
+		if [[ -x "${candidate_path}" ]]; then
+			selected_path="${candidate_path}"
+			break;
+		fi
+	done
+
+	if [[ -z "${selected_path}" ]]; then
+		die "Could not find (suitable) Zig at \"${base_path}\""
+	fi
+
+	export ZIG_EXE="${selected_path}"
+	export ZIG_VER="${selected_ver}"
+	# Sanity check, comment from upstream:
+	# > Check libc++ linkage to make sure Zig was built correctly,
+	# > but only for "env" and "version" to avoid affecting the
+	# > startup time for build-critical commands
+	# > (check takes about ~10 μs)
+	"${ZIG_EXE}" version > /dev/null \
+		|| die "Sanity check failed for \"${ZIG_EXE}\""
+}
+
+# @FUNCTION: zig-toolchain_setup
+# @DESCRIPTION:
+# Checks if running Linux kernel version is supported by Zig.
+# Populates ZIG_TARGET, ZIG_CPU, ZIG_EXE and ZIG_VER environment
+# variables with detected values, or, if user set them already,
+# leaves as-is.
+zig-toolchain_setup() {
+	# Should be first because it sets ZIG_VER which might be used
+	# in the future when setting ZIG_TARGET and ZIG_CPU variables
+	# for incompatible versions.
+	if [[ -z "${ZIG_EXE}" ]]; then
+		zig-toolchain_find_installation
+	fi
+
+	if [[ -z ${ZIG_TARGET} ]]; then
+		if tc-is-cross-compiler; then
+			export ZIG_TARGET=$(zig-toolchain_get_target ${CHOST})
+		else
+			export ZIG_TARGET=native
+		fi
+	fi
+
+	if [[ -z ${ZIG_CPU} ]]; then
+		if tc-is-cross-compiler; then
+			export ZIG_CPU=$(zig-toolchain_get_cpu ${CHOST})
+		else
+			export ZIG_CPU=native
+		fi
+	fi
```

Comment:
Sounds like you could move the `tc-is-cross-compiler` check outside, it's probably more expensive and I think it would look more consistent that way.

Reply:
Like this?

```bash
if tc-is-cross-compiler; then
	if [[ -z ${ZIG_TARGET} ]]; then
		export ZIG_TARGET=$(zig-toolchain_get_target ${CHOST})

	if [[ -z ${ZIG_CPU} ]]; then
		export ZIG_CPU=$(zig-toolchain_get_cpu ${CHOST})
else
	if [[ -z ${ZIG_TARGET} ]]; then
		export ZIG_TARGET=native

	if [[ -z ${ZIG_CPU} ]]; then
		export ZIG_CPU=native
fi
```

Reply:
Yeah. Perhaps you could collapse it more into something like `: "${ZIG_CPU:=...}"`, and then do a common `export ZIG_CPU ZIG_TARGET` below.

Reply:
> Yeah. Perhaps you could collapse it more into something like `: "${ZIG_CPU:=...}"`, and then do a common `export ZIG_CPU ZIG_TARGET` below.

Like this?

```bash
if tc-is-cross-compiler; then
	: "${ZIG_TARGET:=$(zig-toolchain_get_target "${CHOST}")}"
	: "${ZIG_CPU:=$(zig-toolchain_get_cpu "${CHOST}")}"
else
	: "${ZIG_TARGET:=native}"
	: "${ZIG_CPU:=native}"
fi

export ZIG_TARGET ZIG_CPU
```

Reply:
Yep. Though I'd do `ZIG_CPU` first, to sort it lexically.

Reply:
d45c79cffb21...20659c55aec3